### PR TITLE
SI-7852 for GenBCode

### PR DESCRIPTION
**Pull request by lrytz**
Avoid null checks for "someLiteral".== and SomeModule.==. This has
been implemented in GenICode in #2954.

Introduces a trait to share code between GenICode and GenBCode. This
is just a start, more such refactorings will come quite certainly.

Test already exists. The new `scala-nightly-genbcode-2.1X.x` nightly build
runs all tests on the new backend (that's how the difference was found).

### CODE REVIEW

**Comment by lrytz** [comment]
Review by @retronym

**Comment by retronym** [comment]
LGTM. Here's a the [build failure](https://scala-webapps.epfl.ch/jenkins/view/2.11.x/job/scala-nightly-genbcode-2.11.x/1/console) for the record.

**Comment by retronym** [comment]
As a followup, you probably want to find the spots where I've already shared code between the backends, by having one call the other, and move that into the new common trait.

**Comment by lrytz** [comment]
Right, one could also call `genBCode.isNull` in GenICode (or vice versa). I can change things if that is the preferred approach. Can you give me an example of such a call that you introduced?

**Comment by retronym** [comment]
i think your way is cleaner. 

I did this most recently in the #3711.

Before that:

```
git log -i --grep duplication -- src/compiler/scala/tools/nsc/backend/
commit c7a5aab0d34243bfec9de2882594927b460b1ec9
Author: Jason Zaugg <jzaugg@gmail.com>
Date:   Sat Feb 15 14:30:50 2014 +0100

    SI-3452 Reduce low-hanging duplication bewteen GenASM and GenBCode

commit 6e2cadb8bd42e0f4a6c777b11ff71f92cda368e8
Author: Jason Zaugg <jzaugg@gmail.com>
Date:   Mon Sep 16 12:55:10 2013 +0200

    SI-7847 Static forwarders for case apply/unapply

    These were excluded in f901816b3f because at the time they
    were compiler fiction and translated to calls to the case
    class constructor or accessors.

    But since they are now bona-fide methods (albeit still occasionally
    bypassed as an optimization), we can expose them conveniently to our
    Java brethren.

    The cut-and-pastiness of GenBCode starts to hinder maintenance.
    Here's a report of further duplication that we have to fix up
    post haste: https://gist.github.com/retronym/6334389

commit 7d83be218473f08ab560c937318faadce2617d95
Author: Paul Phillips <paulp@improving.org>
Date:   Sun Aug 25 09:22:47 2013 -0700

    Logging cleanup.

    Reduced the amount of extraneous logging noise at the
    default logging level.

    Was brought to my usual crashing halt by the discovery of identical
    logging statements throughout GenASM and elsewhere. I'm supposing
    the reason people so grossly underestimate the cost of such duplication
    is that most of the effects are in things which don't happen, aka
    "silent evidence".

    An example of a thing which isn't happening is the remainder of
    this commit, which exists only in parallel universes.
```

**Comment by gkossakowski** [comment]
Would it be ok to use composition + imports instead of inheritance for code sharing?

``` scala
final class BCodeICodeCommon(global: Global) {
  import global._
  // the common code
}
```

and then

``` scala
abstract class GenICode extends SubComponent {
  private val bCodeICodeCommon = new BCodeICodeCommon(global)
  import bCodeICodeCommon._
} 
```

This would help incremental compilation. If you use inheritance and add something to `BCodeICodeCommon` then trait/class that inherits it directly or indirectly has to be recompiled. If you use composition it's most likely that nothing apart from `BCodeICodeCommon` will get recompiled.

**Comment by lrytz** [comment]
Sure, that's fine with me.

**Comment by gkossakowski** [comment]
Actually, I realized now that you'll need:

``` scala
final class BCodeICodeCommon[ThisGlobal <: Global](global: ThisGlobal)
```

in order to push through the path-dependent types.

**Comment by lrytz** [comment]
@gkossakowski ptal

File: src/compiler/scala/tools/nsc/backend/icode/GenICode.scala [hunk 1]
```diff
@@ -30,6 +29,9 @@ abstract class GenICode extends SubComponent  {
   }
   import platform.isMaybeBoxed
 
+  private val bCodeICodeCommon: jvm.BCodeICodeCommon[GenICode.this.global.type] = new jvm.BCodeICodeCommon(global)
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/backend/icode/GenICode.scala]: Isn't `global.type` enough?

**Comment by lrytz** [comment]
done, thanks

**Comment by gkossakowski** [comment]
LGTM

**Pull request merged**